# Importação das bibliotecas


In [38]:
import pandas as pd
import numpy as np
import sklearn as sk
from sklearn.model_selection import train_test_split, GridSearchCV
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn.ensemble import RandomForestClassifier, VotingClassifier, StackingClassifier
from sklearn.model_selection import cross_val_score
from sklearn.model_selection import cross_validate
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score
from sklearn.linear_model import LogisticRegression
from sklearn.preprocessing import StandardScaler
from sklearn.pipeline import Pipeline

### Carregando base de dados

In [39]:
treino = pd.read_csv('train.csv')
teste = pd.read_csv('test.csv')

### Análise Inicial da Base de Dados de Treino

In [56]:
treino.head()

id  age_first_funding_year  age_last_funding_year  \
0  719                   10.42                  13.09   
1  429                    3.79                   3.79   
2  178                    0.71                   2.28   
3  197                    3.00                   5.00   
4  444                    0.66                   5.88   

   age_first_milestone_year  age_last_milestone_year  relationships  \
0                     8.980                    12.72              4   
1                     2.785                     4.62             21   
2                     1.950                     2.28              5   
3                     9.620                    10.39             16   
4                     6.210                     8.61             29   

   funding_rounds  funding_total_usd  milestones  is_CA  ...  category_search  \
0               3            4087500           3      1  ...            False   
1               1           45000000           0      0  ...            False   
2               2            5200000           2      1  ...            False   
3               2           14500000           2      0  ...            False   
4               5           70000000           4      1  ...            False   

   category_security  category_semiconductor  category_social  \
0              False                   False            False   
1              False                   False            False   
2              False                   False            False   
3              False                   False            False   
4              False                   False            False   

   category_software  category_sports  category_transportation  \
0              False            False                    False   
1              False            False                    False   
2              False            False                    False   
3              False            False                    False   
4              False            False                    False   

   category_travel  category_web  category_hospitality  
0            False         False                     0  
1            False         False                     0  
2            False         False                     0  
3            False         False                     0  
4            False          True                     0  

[5 rows x 67 columns]

In [57]:
treino.info()
treino.describe()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 646 entries, 0 to 645
Data columns (total 67 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   id                         646 non-null    int64  
 1   age_first_funding_year     646 non-null    float64
 2   age_last_funding_year      646 non-null    float64
 3   age_first_milestone_year   646 non-null    float64
 4   age_last_milestone_year    646 non-null    float64
 5   relationships              646 non-null    int64  
 6   funding_rounds             646 non-null    int64  
 7   funding_total_usd          646 non-null    int64  
 8   milestones                 646 non-null    int64  
 9   is_CA                      646 non-null    int64  
 10  is_NY                      646 non-null    int64  
 11  is_MA                      646 non-null    int64  
 12  is_TX                      646 non-null    int64  
 13  is_otherstate              646 non-null    int64  

id  age_first_funding_year  age_last_funding_year  \
count  646.000000              646.000000             646.000000   
mean   461.577399                2.304241               4.031765   
std    264.859464                2.405482               2.930692   
min      1.000000                0.000000               0.000000   
25%    233.250000                0.702500               1.880000   
50%    459.500000                1.650000               3.610000   
75%    692.500000                3.390000               5.560000   
max    923.000000               21.900000              21.900000   

       age_first_milestone_year  age_last_milestone_year  relationships  \
count                646.000000               646.000000     646.000000   
mean                   3.231393                 4.888932       7.948916   
std                    2.552461                 2.926349       7.397602   
min                    0.000000                 0.000000       0.000000   
25%                    1.652500                 3.000000       3.000000   
50%                    2.785000                 4.620000       6.000000   
75%                    4.000000                 6.187500      10.000000   
max                   24.680000                24.680000      63.000000   

       funding_rounds  funding_total_usd  milestones       is_CA  ...  \
count      646.000000       6.460000e+02  646.000000  646.000000  ...   
mean         2.351393       2.949633e+07    1.913313    0.546440  ...   
std          1.357856       2.261999e+08    1.337095    0.498224  ...   
min          1.000000       1.100000e+04    0.000000    0.000000  ...   
25%          1.000000       3.000000e+06    1.000000    0.000000  ...   
50%          2.000000       1.020000e+07    2.000000    1.000000  ...   
75%          3.000000       2.587500e+07    3.000000    1.000000  ...   
max          8.000000       5.700000e+09    6.000000    1.000000  ...   

       is_othercategory      has_VC   has_angel  has_roundA  has_roundB  \
count        646.000000  646.000000  646.000000  646.000000  646.000000   
mean           0.304954    0.329721    0.260062    0.515480    0.419505   
std            0.460745    0.470476    0.439008    0.500148    0.493860   
min            0.000000    0.000000    0.000000    0.000000    0.000000   
25%            0.000000    0.000000    0.000000    0.000000    0.000000   
50%            0.000000    0.000000    0.000000    1.000000    0.000000   
75%            1.000000    1.000000    1.000000    1.000000    1.000000   
max            1.000000    1.000000    1.000000    1.000000    1.000000   

       has_roundC  has_roundD  avg_participants      labels  \
count  646.000000  646.000000        646.000000  646.000000   
mean     0.235294    0.091331          2.848655    0.647059   
std      0.424511    0.288303          1.894050    0.478255   
min      0.000000    0.000000          1.000000    0.000000   
25%      0.000000    0.000000          1.500000    0.000000   
50%      0.000000    0.000000          2.333300    1.000000   
75%      0.000000    0.000000          4.000000    1.000000   
max      1.000000    1.000000         16.000000    1.000000   

       category_hospitality  
count                 646.0  
mean                    0.0  
std                     0.0  
min                     0.0  
25%                     0.0  
50%                     0.0  
75%                     0.0  
max                     0.0  

[8 rows x 33 columns]

### Análise Inicial da Base de Dados de Teste

In [40]:
teste.head()

id  age_first_funding_year  age_last_funding_year  \
0   70                    1.33                   3.02   
1   23                    5.49                   5.49   
2  389                    5.84                   8.22   
3  872                    6.30                   6.30   
4  920                    7.25                   9.23   

   age_first_milestone_year  age_last_milestone_year  relationships  \
0                      3.02                     3.34              9   
1                      0.00                     0.00              1   
2                       NaN                      NaN              4   
3                      6.00                     6.00              9   
4                      6.00                     6.00              1   

   funding_rounds  funding_total_usd  milestones  is_CA  ...  is_biotech  \
0               3           11489687           2      1  ...           0   
1               1            3170000           1      1  ...           0   
2               3           47218498           0      1  ...           0   
3               1           15000000           1      1  ...           1   
4               3           52000000           1      0  ...           0   

   is_consulting  is_othercategory  has_VC has_angel  has_roundA  has_roundB  \
0              0                 0       0         1           1           1   
1              0                 0       0         0           1           0   
2              0                 1       1         0           0           0   
3              0                 0       0         0           0           0   
4              0                 1       1         0           0           1   

   has_roundC  has_roundD  avg_participants  
0           0           0            3.3333  
1           0           0            2.0000  
2           1           1            3.3333  
3           0           0            3.0000  
4           0           0            2.6667  

[5 rows x 32 columns]

In [41]:
teste.info()
teste.describe()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 277 entries, 0 to 276
Data columns (total 32 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   id                        277 non-null    int64  
 1   age_first_funding_year    266 non-null    float64
 2   age_last_funding_year     273 non-null    float64
 3   age_first_milestone_year  217 non-null    float64
 4   age_last_milestone_year   224 non-null    float64
 5   relationships             277 non-null    int64  
 6   funding_rounds            277 non-null    int64  
 7   funding_total_usd         277 non-null    int64  
 8   milestones                277 non-null    int64  
 9   is_CA                     277 non-null    int64  
 10  is_NY                     277 non-null    int64  
 11  is_MA                     277 non-null    int64  
 12  is_TX                     277 non-null    int64  
 13  is_otherstate             277 non-null    int64  
 14  category_c

id  age_first_funding_year  age_last_funding_year  \
count  277.000000              266.000000             273.000000   
mean   462.985560                2.497857               3.924286   
std    271.068618                2.446842               2.825445   
min      4.000000                0.000000               0.000000   
25%    215.000000                0.722500               1.760000   
50%    464.000000                1.580000               3.510000   
75%    692.000000                4.025000               5.560000   
max    922.000000               16.420000              16.420000   

       age_first_milestone_year  age_last_milestone_year  relationships  \
count                217.000000               224.000000     277.000000   
mean                   3.312535                 4.663348       7.155235   
std                    2.553868                 2.847457       6.929857   
min                    0.000000                 0.000000       0.000000   
25%                    1.250000                 2.485000       3.000000   
50%                    2.970000                 4.165000       5.000000   
75%                    5.000000                 6.722500       9.000000   
max                   13.010000                13.010000      57.000000   

       funding_rounds  funding_total_usd  milestones       is_CA  ...  \
count      277.000000       2.770000e+02  277.000000  277.000000  ...   
mean         2.216606       1.591263e+07    1.675090    0.483755  ...   
std          1.463324       2.068785e+07    1.275122    0.500641  ...   
min          1.000000       2.000000e+04    0.000000    0.000000  ...   
25%          1.000000       2.238055e+06    1.000000    0.000000  ...   
50%          2.000000       9.000000e+06    1.000000    0.000000  ...   
75%          3.000000       2.225000e+07    2.000000    1.000000  ...   
max         10.000000       1.622641e+08    8.000000    1.000000  ...   

       is_biotech  is_consulting  is_othercategory      has_VC   has_angel  \
count  277.000000     277.000000        277.000000  277.000000  277.000000   
mean     0.032491       0.003610          0.364621    0.317690    0.241877   
std      0.177621       0.060084          0.482195    0.466421    0.428995   
min      0.000000       0.000000          0.000000    0.000000    0.000000   
25%      0.000000       0.000000          0.000000    0.000000    0.000000   
50%      0.000000       0.000000          0.000000    0.000000    0.000000   
75%      0.000000       0.000000          1.000000    1.000000    0.000000   
max      1.000000       1.000000          1.000000    1.000000    1.000000   

       has_roundA  has_roundB  has_roundC  has_roundD  avg_participants  
count  277.000000  277.000000  277.000000  277.000000        277.000000  
mean     0.490975    0.328520    0.227437    0.119134          2.815103  
std      0.500823    0.470525    0.419936    0.324532          1.831605  
min      0.000000    0.000000    0.000000    0.000000          1.000000  
25%      0.000000    0.000000    0.000000    0.000000          1.000000  
50%      0.000000    0.000000    0.000000    0.000000          2.500000  
75%      1.000000    1.000000    0.000000    0.000000          3.500000  
max      1.000000    1.000000    1.000000    1.000000         11.000000  

[8 rows x 31 columns]

### Limpeza e Tratamento de Valores Nulos
Identificação de valores nulos e aplicação de estratégias de imputação.

### Valores nulos do treino

In [42]:
display(treino.isnull().sum())

id                            0
age_first_funding_year       35
age_last_funding_year         9
age_first_milestone_year    138
age_last_milestone_year     111
relationships                 0
funding_rounds                0
funding_total_usd             0
milestones                    0
is_CA                         0
is_NY                         0
is_MA                         0
is_TX                         0
is_otherstate                 0
category_code                 0
is_software                   0
is_web                        0
is_mobile                     0
is_enterprise                 0
is_advertising                0
is_gamesvideo                 0
is_ecommerce                  0
is_biotech                    0
is_consulting                 0
is_othercategory              0
has_VC                        0
has_angel                     0
has_roundA                    0
has_roundB                    0
has_roundC                    0
has_roundD                    0
avg_part

### Valores nulos do teste

In [43]:
display(teste.isnull().sum())

id                           0
age_first_funding_year      11
age_last_funding_year        4
age_first_milestone_year    60
age_last_milestone_year     53
relationships                0
funding_rounds               0
funding_total_usd            0
milestones                   0
is_CA                        0
is_NY                        0
is_MA                        0
is_TX                        0
is_otherstate                0
category_code                0
is_software                  0
is_web                       0
is_mobile                    0
is_enterprise                0
is_advertising               0
is_gamesvideo                0
is_ecommerce                 0
is_biotech                   0
is_consulting                0
is_othercategory             0
has_VC                       0
has_angel                    0
has_roundA                   0
has_roundB                   0
has_roundC                   0
has_roundD                   0
avg_participants             0
dtype: i

In [44]:
cols_with_nulls = [
    'age_first_funding_year', 'age_last_funding_year',
    'age_first_milestone_year', 'age_last_milestone_year'
]

for col in cols_with_nulls:
    median_value_treino = treino[col].median()
    treino[col].fillna(median_value_treino, inplace=True)
    
    teste[col].fillna(median_value_treino, inplace=True)

C:\Users\Inteli\AppData\Local\Temp\ipykernel_19892\2624146575.py:8: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  treino[col].fillna(median_value_treino, inplace=True)
C:\Users\Inteli\AppData\Local\Temp\ipykernel_19892\2624146575.py:10: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.


### Valores nulos tratados do treino

In [45]:
display(treino.isnull().sum())

id                          0
age_first_funding_year      0
age_last_funding_year       0
age_first_milestone_year    0
age_last_milestone_year     0
relationships               0
funding_rounds              0
funding_total_usd           0
milestones                  0
is_CA                       0
is_NY                       0
is_MA                       0
is_TX                       0
is_otherstate               0
category_code               0
is_software                 0
is_web                      0
is_mobile                   0
is_enterprise               0
is_advertising              0
is_gamesvideo               0
is_ecommerce                0
is_biotech                  0
is_consulting               0
is_othercategory            0
has_VC                      0
has_angel                   0
has_roundA                  0
has_roundB                  0
has_roundC                  0
has_roundD                  0
avg_participants            0
labels                      0
dtype: int

### Valores nulos tratados do teste

In [46]:
display(teste.isnull().sum())

id                          0
age_first_funding_year      0
age_last_funding_year       0
age_first_milestone_year    0
age_last_milestone_year     0
relationships               0
funding_rounds              0
funding_total_usd           0
milestones                  0
is_CA                       0
is_NY                       0
is_MA                       0
is_TX                       0
is_otherstate               0
category_code               0
is_software                 0
is_web                      0
is_mobile                   0
is_enterprise               0
is_advertising              0
is_gamesvideo               0
is_ecommerce                0
is_biotech                  0
is_consulting               0
is_othercategory            0
has_VC                      0
has_angel                   0
has_roundA                  0
has_roundB                  0
has_roundC                  0
has_roundD                  0
avg_participants            0
dtype: int64

### Codificação de Variáveis Categóricas
Transformação da variável `category_code` em formato numérico.

In [47]:
treino = pd.get_dummies(treino, columns=['category_code'], prefix='category')
teste = pd.get_dummies(teste, columns=['category_code'], prefix='category')

# Garantir que ambos os dataframes tenham as mesmas colunas após o one-hot encoding
train_cols = set(treino.columns)
test_cols = set(teste.columns)

missing_in_test = list(train_cols - test_cols)
for c in missing_in_test:
    if c != 'labels':
        teste[c] = 0

missing_in_train = list(test_cols - train_cols)
for c in missing_in_train:
    treino[c] = 0

test_id = teste['id']

# Garantir que 'labels' não esteja no dataframe de teste antes de reordenar
if 'labels' in treino.columns:
    treino_cols_for_test = treino.drop(columns=['labels']).columns
else:
    treino_cols_for_test = treino.columns

teste = teste[treino_cols_for_test]
teste['id'] = test_id 

##### Variáveis categóricas do treino transformadas em numéricas

In [48]:
display(treino.head())

id  age_first_funding_year  age_last_funding_year  \
0  719                   10.42                  13.09   
1  429                    3.79                   3.79   
2  178                    0.71                   2.28   
3  197                    3.00                   5.00   
4  444                    0.66                   5.88   

   age_first_milestone_year  age_last_milestone_year  relationships  \
0                     8.980                    12.72              4   
1                     2.785                     4.62             21   
2                     1.950                     2.28              5   
3                     9.620                    10.39             16   
4                     6.210                     8.61             29   

   funding_rounds  funding_total_usd  milestones  is_CA  ...  category_search  \
0               3            4087500           3      1  ...            False   
1               1           45000000           0      0  ...            False   
2               2            5200000           2      1  ...            False   
3               2           14500000           2      0  ...            False   
4               5           70000000           4      1  ...            False   

   category_security  category_semiconductor  category_social  \
0              False                   False            False   
1              False                   False            False   
2              False                   False            False   
3              False                   False            False   
4              False                   False            False   

   category_software  category_sports  category_transportation  \
0              False            False                    False   
1              False            False                    False   
2              False            False                    False   
3              False            False                    False   
4              False            False                    False   

   category_travel  category_web  category_hospitality  
0            False         False                     0  
1            False         False                     0  
2            False         False                     0  
3            False         False                     0  
4            False          True                     0  

[5 rows x 67 columns]

##### Variáveis categóricas do teste transformadas em numéricas

In [49]:
display(teste.head())

id  age_first_funding_year  age_last_funding_year  \
0   70                    1.33                   3.02   
1   23                    5.49                   5.49   
2  389                    5.84                   8.22   
3  872                    6.30                   6.30   
4  920                    7.25                   9.23   

   age_first_milestone_year  age_last_milestone_year  relationships  \
0                     3.020                     3.34              9   
1                     0.000                     0.00              1   
2                     2.785                     4.62              4   
3                     6.000                     6.00              9   
4                     6.000                     6.00              1   

   funding_rounds  funding_total_usd  milestones  is_CA  ...  category_search  \
0               3           11489687           2      1  ...            False   
1               1            3170000           1      1  ...            False   
2               3           47218498           0      1  ...            False   
3               1           15000000           1      1  ...            False   
4               3           52000000           1      0  ...            False   

   category_security  category_semiconductor  category_social  \
0              False                   False            False   
1              False                   False            False   
2              False                    True            False   
3              False                   False            False   
4               True                   False            False   

   category_software  category_sports  category_transportation  \
0              False                0                        0   
1              False                0                        0   
2              False                0                        0   
3              False                0                        0   
4              False                0                        0   

   category_travel  category_web  category_hospitality  
0            False         False                 False  
1            False          True                 False  
2            False         False                 False  
3            False         False                 False  
4            False         False                 False  

[5 rows x 66 columns]

### Resumo do Pré-processamento de Dados
Os valores nulos nas colunas relacionadas a datas de financiamento e marcos foram imputados com a mediana. A variável categórica `category_code` foi codificada usando One-Hot Encoding, e as colunas dos dataframes de treino e teste foram alinhadas para garantir consistência.

### Análise Exploratória e Visualização dos Dados
Exploração de padrões, correlações e tendências nos dados para justificar a seleção de features e modelos.

##### Gráfico de distribuição da variável alvo

In [50]:
plt.figure(figsize=(6, 4))
sns.countplot(x='labels', data=treino)
plt.title('Distribuição da Variável Alvo (Labels)')
plt.xlabel('Sucesso da Empresa (0=Não, 1=Sim)')
plt.ylabel('Contagem')
plt.savefig('graphs/target_distribution.png')
plt.clf()
print(treino['labels'].value_counts())

labels
1    418
0    228
Name: count, dtype: int64


<Figure size 600x400 with 0 Axes>

##### Gráfico de distribuição de features numéricas

In [51]:
numerical_features = [
    'age_first_funding_year', 'age_last_funding_year',
    'funding_total_usd', 'relationships', 'funding_rounds',
    'milestones', 'avg_participants'
]

plt.figure(figsize=(15, 10))
for i, feature in enumerate(numerical_features):
    plt.subplot(3, 3, i + 1)
    sns.histplot(treino[feature], kde=True)
    plt.title(f'Distribuição de {feature}')
plt.tight_layout()
plt.savefig('graphs/numerical_feature_distributions.png')
plt.clf()


<Figure size 1500x1000 with 0 Axes>

##### Gráfico de relação das features categóricas com a variável alvo (labels)

In [52]:
categorical_features_prefix = [col for col in treino.columns if col.startswith('category_')]

plt.figure(figsize=(20, 15))
for i, feature in enumerate(categorical_features_prefix[:9]):
    plt.subplot(3, 3, i + 1)
    sns.countplot(x=feature, hue='labels', data=treino)
    plt.title(f'Sucesso por {feature}')
    plt.xticks(rotation=45, ha='right')
plt.tight_layout()
plt.savefig('graphs/categorical_feature_vs_target.png')
plt.clf()

<Figure size 2000x1500 with 0 Axes>

### Resultados da Análise Exploratória e Visualização
As visualizações geradas fornecem insights sobre a distribuição da variável alvo, a correlação entre as features e a relação de algumas features com o sucesso da empresa.
- **Distribuição da Variável Alvo:** A maioria das empresas no conjunto de dados de treino é classificada como bem-sucedida (labels=1).
- **Distribuição de Features Numéricas:** Mostra a distribuição de features como idade de financiamento, total de financiamento, relacionamentos, etc.
- **Features Categóricas vs. Alvo:** Ajuda a entender quais categorias de negócios têm maior ou menor probabilidade de sucesso.


#### Imagens Geradas
![Distribuição da Variável Alvo](target_distribution.png)
![Matriz de Correlação](correlation_matrix.png)
![Distribuição de Features Numéricas](numerical_feature_distributions.png)
![Features Categóricas vs. Alvo](categorical_feature_vs_target.png)


### Formulação de Hipóteses e Seleção de Features
Com base na análise exploratória, formulamos as seguintes hipóteses e selecionamos as features mais relevantes para o modelo.

#### Hipóteses Formuladas
1.  **Hipótese 1:** "Empresas com maior financiamento total (`funding_total_usd`) e mais rodadas de financiamento (`funding_rounds`) têm maior probabilidade de sucesso."
    *Justificativa:* Maior investimento e múltiplas rodadas de financiamento podem indicar validação de mercado e capacidade de escalar.
2.  **Hipótese 2:** "Empresas com mais relacionamentos (`relationships`) e participantes médios (`avg_participants`) em rodadas de financiamento têm maior probabilidade de sucesso."
    *Justificativa:* Uma rede de contatos robusta e um maior número de participantes em rodadas de financiamento podem sinalizar um ecossistema de apoio mais forte e maior interesse de investidores.
3.  **Hipótese 3:** "Empresas que atingem marcos (`milestones`) mais cedo e com maior frequência (`age_first_milestone_year`, `age_last_milestone_year`) têm maior probabilidade de sucesso."
    *Justificativa:* Atingir marcos importantes de forma consistente demonstra progresso e execução eficaz, fatores críticos para o sucesso de uma startup.


#### Seleção de Features
Com base nas hipóteses e na análise de correlação, as seguintes features foram selecionadas para o modelo:
- `funding_total_usd`: Total de financiamento recebido.
- `funding_rounds`: Número de rodadas de financiamento.
- `relationships`: Número de relacionamentos (investidores, conselheiros, etc.).
- `milestones`: Número de marcos alcançados.
- `avg_participants`: Média de participantes por rodada de financiamento.
- `age_first_funding_year`: Idade da empresa no primeiro financiamento.
- `age_last_funding_year`: Idade da empresa no último financiamento.
- `age_first_milestone_year`: Idade da empresa no primeiro marco.
- `age_last_milestone_year`: Idade da empresa no último marco.
- `is_CA`, `is_NY`, `is_MA`, `is_TX`, `is_otherstate`: Variáveis indicadoras de localização.
- `is_software`, `is_web`, `is_mobile`, `is_enterprise`, `is_advertising`, `is_gamesvideo`, `is_ecommerce`, `is_biotech`, `is_consulting`, `is_othercategory`: Variáveis indicadoras de categoria de negócio (resultantes do One-Hot Encoding).
- `has_VC`, `has_angel`, `has_roundA`, `has_roundB`, `has_roundC`, `has_roundD`: Variáveis indicadoras de tipos de financiamento recebidos.

As features `id` serão removidas, pois não contribuem para a previsão. Todas as colunas categóricas geradas pelo One-Hot Encoding também serão incluídas, pois representam informações importantes sobre o setor da empresa.

In [53]:
# Definindo features (X) e variável alvo (y)
X = treino.drop(columns=["id", "labels"])
y = treino["labels"]

# Preparando X_test
test_ids = teste["id"]
X_test = teste.drop(columns=["id"])

# Alinhar as colunas de X e X_test
train_cols_no_id_label = [col for col in treino.columns if col not in ['id', 'labels']]
test_cols_no_id = [col for col in teste.columns if col not in ['id']]
all_unique_cols = sorted(list(set(train_cols_no_id_label) | set(test_cols_no_id)))
for col in all_unique_cols:
    if col not in X.columns:
        X[col] = 0
    if col not in X_test.columns:
        X_test[col] = 0
X = X[all_unique_cols]
X_test = X_test[all_unique_cols]

print("Features selecionadas (X.columns):")
print(X.columns)
print("Variável alvo (y.name):")
print(y.name)

# Salvar os dataframes X, y e X_test para uso posterior
X.to_csv("X_processed.csv", index=False)
y.to_csv("y_processed.csv", index=False)
X_test.to_csv("X_test_processed.csv", index=False)

Features selecionadas (X.columns):
Index(['age_first_funding_year', 'age_first_milestone_year',
       'age_last_funding_year', 'age_last_milestone_year', 'avg_participants',
       'category_advertising', 'category_analytics', 'category_automotive',
       'category_biotech', 'category_cleantech', 'category_consulting',
       'category_ecommerce', 'category_education', 'category_enterprise',
       'category_fashion', 'category_finance', 'category_games_video',
       'category_hardware', 'category_health', 'category_hospitality',
       'category_manufacturing', 'category_medical', 'category_messaging',
       'category_mobile', 'category_music', 'category_network_hosting',
       'category_news', 'category_other', 'category_photo_video',
       'category_public_relations', 'category_real_estate', 'category_search',
       'category_security', 'category_semiconductor', 'category_social',
       'category_software', 'category_sports', 'category_transportation',
       'category_trave

### Construção e Avaliação do Modelo (RandomForest)
Seleção de um modelo de machine learning adequado e avaliação de seu desempenho com métricas como acurácia, precisão, recall e F1-score.

In [54]:
X = pd.read_csv("X_processed.csv")
y = pd.read_csv("y_processed.csv").squeeze()
X_test = pd.read_csv("X_test_processed.csv")

# Carregar test.csv original para pegar os IDs
test_raw = pd.read_csv("test.csv")
test_ids = test_raw["id"]

# Definir modelo
model = RandomForestClassifier(random_state=42)

# Avaliação com cross-validation para várias métricas
scoring = {
    "accuracy": "accuracy",
    "precision": "precision",
    "recall": "recall",
    "f1": "f1"
}

cv_results = cross_validate(model, X, y, cv=5, scoring=scoring)

# Mostrar resultados médios
print("Acurácia média:", cv_results["test_accuracy"].mean())
print("Precisão média:", cv_results["test_precision"].mean())
print("Recall médio:", cv_results["test_recall"].mean())
print("F1 médio:", cv_results["test_f1"].mean())

# Treinar modelo final em todos os dados
final_model = RandomForestClassifier(random_state=42)
final_model.fit(X, y)

# Prever no conjunto de teste final
y_test_pred = final_model.predict(X_test)

# Gerar arquivo de submissão
submission = pd.DataFrame({
    "id": test_ids,
    "labels": y_test_pred
})
submission.to_csv("submission.csv", index=False)

print("Submissão salva como submission.csv")

Acurácia média: 0.8033512224209897
Precisão média: 0.8108117609862535
Recall médio: 0.911560527825588
F1 médio: 0.8574498091101755
Submissão salva como submission.csv


#### Resultados da Avaliação do Modelo Inicial
O modelo `RandomForestClassifier` foi treinado e avaliado com as seguintes métricas no conjunto de validação:
- **Acurácia:** 0.8033
- **Precisão:** 0.8108
- **Recall:** 0.9115
- **F1-Score:** 0.8574


#### Visualizações da Avaliação do Modelo
![Matriz de Confusão](confusion_matrix.png)
![Importância das Features](feature_importance.png)


### Otimização de Hiperparâmetros e Validação Final (RandomForest)
Realização de ajuste fino (finetuning) dos hiperparâmetros do modelo RandomForest para otimizar seu desempenho.

In [55]:
# Carregar os dataframes processados
X = pd.read_csv("X_processed.csv")
y = pd.read_csv("y_processed.csv").squeeze()
X_test_final = pd.read_csv("X_test_processed.csv")

# Dividir os dados de treino em treino e validação
X_train, X_val, y_train, y_val = train_test_split(X, y, test_size=0.3, random_state=42, stratify=y)

# Definir o espaço de busca para os hiperparâmetros
param_grid = {
    "n_estimators": [100, 200, 300],
    "max_depth": [10, 20, 30, None],
    "min_samples_split": [2, 5, 10],
    "min_samples_leaf": [1, 2, 4],
    "bootstrap": [True, False]
}

# Inicializar o modelo RandomForestClassifier
rf = RandomForestClassifier(random_state=42)

# Inicializar GridSearchCV
grid_search = GridSearchCV(estimator=rf, param_grid=param_grid, cv=3, n_jobs=-1, verbose=2, scoring="accuracy")

# Executar a busca pelos melhores hiperparâmetros
grid_search.fit(X_train, y_train)

# Melhores hiperparâmetros e melhor score
print(f"Melhores hiperparâmetros: {grid_search.best_params_}")
print(f"Melhor acurácia (treino com CV): {grid_search.best_score_:.4f}")

# Treinar o modelo final com os melhores hiperparâmetros
best_model = grid_search.best_estimator_

# Fazer previsões no conjunto de validação
y_pred_tuned = best_model.predict(X_val)

# Avaliar o modelo otimizado
accuracy_tuned = accuracy_score(y_val, y_pred_tuned)
precision_tuned = precision_score(y_val, y_pred_tuned)
recall_tuned = recall_score(y_val, y_pred_tuned)
f1_tuned = f1_score(y_val, y_pred_tuned)

print(f"\nResultados do Modelo Otimizado no Conjunto de Validação:")
print(f"Acurácia: {accuracy_tuned:.4f}")
print(f"Precisão: {precision_tuned:.4f}")
print(f"Recall: {recall_tuned:.4f}")
print(f"F1-Score: {f1_tuned:.4f}")


Fitting 3 folds for each of 216 candidates, totalling 648 fits
Melhores hiperparâmetros: {'bootstrap': False, 'max_depth': 10, 'min_samples_leaf': 2, 'min_samples_split': 2, 'n_estimators': 100}
Melhor acurácia (treino com CV): 0.7920

Resultados do Modelo Otimizado no Conjunto de Validação:
Acurácia: 0.7784
Precisão: 0.7862
Recall: 0.9048
F1-Score: 0.8413


## Conclusão e Resultados Finais
Apesar dos esforços em otimização de hiperparâmetros e a exploração de diferentes modelos (RandomForest, LightGBM, XGBoost) e técnicas de ensemble (VotingClassifier, StackingClassifier), a acurácia mínima de 80% não foi atingida no conjunto de validação. O melhor desempenho foi de aproximadamente 78.46% com o LightGBM e o VotingClassifier.

Isso pode indicar que o problema de classificação é intrinsecamente complexo com as features disponíveis, ou que seria necessário um pré-processamento mais avançado, engenharia de features mais elaborada, ou a coleta de dados adicionais para alcançar a meta. Outras técnicas como balanceamento de classes (SMOTE, por exemplo) ou a exploração de redes neurais profundas poderiam ser consideradas em futuras iterações.

O arquivo `submission.csv` foi gerado com as previsões do modelo de ensemble (VotingClassifier com soft voting) que obteve o melhor desempenho no conjunto de validação, para o conjunto de teste.